### PR TITLE
Integration instructor

### LINKED ISSUE #661: [BUG] Cannot import `create_distiset` directly.
**Describe the bug**
We cannot import `create_distiset` apparently due to a circular import.

**To Reproduce**
Code to reproduce
```python
>>> from distilabel.distiset import create_distiset
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ in <module>:1                                                                                    │
│                                                                                                  │
│ /Users/agus/github_repos/argilla-io/distilabel/src/distilabel/distiset.py:24 in <module>         │
│                                                                                                  │
│    21 from huggingface_hub import DatasetCardData, HfApi                                         │
│    22 from pyarrow.lib import ArrowInvalid                                                       │
│    23                                                                                            │
│ ❱  24 from distilabel.steps.tasks.base import DISTILABEL_METADATA_KEY                            │
│    25 from distilabel.utils.card.dataset_card import (                                           │
│    26 │   DistilabelDatasetCard,                                                                 │
│    27 │   size_categories_parser,                                                                │
│                                                                                                  │
│ /Users/agus/github_repos/argilla-io/distilabel/src/distilabel/steps/__init__.py:18 in <module>   │
│                                                                                                  │
│   15 from distilabel.steps.argilla.preference import PreferenceToArgilla                         │
│   16 from distilabel.steps.argilla.text_generation import TextGenerationToArgilla                │
│   17 from distilabel.steps.base import GeneratorStep, GlobalStep, Step, StepInput                │
│ ❱ 18 from distilabel.steps.combine import CombineColumns                                         │
│   19 from distilabel.steps.decorator import step                                                 │
│   20 from distilabel.steps.deita import DeitaFiltering                                           │
│   21 from distilabel.steps.expand import ExpandColumns                                           │
│                                                                                                  │
│ /Users/agus/github_repos/argilla-io/distilabel/src/distilabel/steps/combine.py:19 in <module>    │
│                                                                                                  │
│   16                                                                                             │
│   17 from typing_extensions import override                                                      │
│   18                                                                                             │
│ ❱ 19 from distilabel.pipeline.utils import combine_dicts                                         │
│   20 from distilabel.steps.base import Step, StepInput                                           │
│   21                                                                                             │
│   22 if TYPE_CHECKING:                                                                           │
│                                                                                                  │
│ /Users/agus/github_repos/argilla-io/distilabel/src/distilabel/pipeline/__init__.py:15 in         │
│ <module>                                                                                         │
│                                                                                                  │
│   12 # See the License for the specific language governing permissions and                       │
│   13 # limitations under the License.                                                            │
│   14                                                                                             │
│ ❱ 15 from distilabel.pipeline.local import Pipeline                                              │
│   16 from distilabel.pipeline.routing_batch_function import (                                    │
│   17 │   routing_batch_function,                                                                 │
│   18 │   sample_n_steps,                                                                         │
│                                                                                                  │
│ /Users/agus/github_repos/argilla-io/distilabel/src/distilabel/pipeline/local.py:25 in <module>   │
│                                                                                                  │
│     22                                                                                           │
│     23 import tblib                                                                              │
│     24                                                                                           │
│ ❱   25 from distilabel.distiset import create_distiset                                           │
│     26 from distilabel.llms.mixins import CudaDevicePlacementMixin                               │
│     27 from distilabel.pipeline.base import (                                                    │
│     28 │   LAST_BATCH_SENT_FLAG,                                                                 │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
ImportError: cannot import name 'create_distiset' from partially initialized module 'distilabel.distiset' (most likely due to a circular import) (/Users/agus/github_repos/argilla-io/distilabel/src/distilabel/distiset.py)
```

**Expected behaviour**

**Screenshots**

**Desktop (please complete the following information):**
 - Package version:
 - Python version:

**Additional context**

### PR DESCRIPTION
**Pull request by plaguss**
## Description

This PR integrates [`instructor`](https://python.useinstructor.com/) for structured generation of LLMs.
It includes support for the following LLMs:
- `openai/azure`, `anthropic`, `litellm`, `mistral`, `cohere` and `groq`.

An example pipeline can be seen in the `/examples/structured_generation_with_instructor.py`. 

The resulting dataset can be seen [here](https://huggingface.co/datasets/distilabel-internal-testing/knowledge_graphs).

Also closes #661

### CODE REVIEW

File: src/distilabel/steps/tasks/structured_outputs/instructor.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    """Wraps the given client with the instructor client for the given framework.
+
+    Args:
+        client: The client to wrap with the instructor client, corresponds to the internal
+            client we wrap on `LLM`, and one of the implemented in `instructor`.
+        mode: One of the `instructor.Mode` values. Defaults to None.
+        framework: The framework corresponding to the client. Defaults to None.
+
+    Raises:
+        ImportError: If instructor is not installed.
```
Comment (gabrielmbmb) [inline src/distilabel/steps/tasks/structured_outputs/instructor.py:115]:
```suggestion
        ImportError: If `instructor` is not installed.
```

File: src/distilabel/steps/tasks/structured_outputs/instructor.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        ValueError: If the mode is not one of the available modes.
+
+    Returns:
+        patched_client: The instructor wrapping the original client to be used for
+            structured generation.
+    """
+    if not importlib.util.find_spec("instructor"):
+        raise ImportError(
+            "instructor is not installed. Please install it using `pip install instructor`."
```
Comment (gabrielmbmb) [inline src/distilabel/steps/tasks/structured_outputs/instructor.py:124]:
```suggestion
            "`instructor` is not installed. Please install it using `pip install instructor`."
```

File: src/distilabel/steps/tasks/structured_outputs/utils.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+# NOTE: The following functions were copied from:
+# https://github.com/pydantic/pydantic/issues/643#issuecomment-1999755873
+# and slightly modified to work with nested models.
+# It would be nice to find the original source of this code to give credit.
+# Other option would be working with this library: https://github.com/c32168/dyntamic
+
+
+def json_schema_to_model(json_schema: Dict[str, Any]) -> Type[BaseModel]:
+    """Converts a JSON schema to a Pydantic BaseModel class.
```
Comment (gabrielmbmb) [inline src/distilabel/steps/tasks/structured_outputs/utils.py:38]:
```suggestion
    """Converts a JSON schema to a `pydantic.BaseModel` class.
```

File: src/distilabel/steps/tasks/structured_outputs/utils.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+    Args:
+        json_schema: The JSON schema to convert.
+
+    Returns:
+        A Pydantic BaseModel class.
```
Comment (gabrielmbmb) [inline src/distilabel/steps/tasks/structured_outputs/utils.py:44]:
```suggestion
        A `pydantic.BaseModel` class.
```

File: src/distilabel/steps/tasks/structured_outputs/utils.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+) -> Any:
+    """Converts a JSON schema property to a Pydantic field definition.
+
+    Args:
+        name: The field name.
+        json_schema: The JSON schema property.
+
+    Returns:
+        A Pydantic field definition.
+    """
```
Comment (gabrielmbmb) [inline src/distilabel/steps/tasks/structured_outputs/utils.py:80]: some missing args in the docstring.

File: src/distilabel/steps/tasks/structured_outputs/utils.py [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+    return create_model(model_name, **field_definitions)
+
+
+def json_schema_to_pydantic_field(
+    name: str,
+    json_schema: Dict[str, Any],
+    required: List[str],
+    defs: Optional[Dict[str, Any]] = None,
+) -> Any:
+    """Converts a JSON schema property to a Pydantic field definition.
```
Comment (gabrielmbmb) [inline src/distilabel/steps/tasks/structured_outputs/utils.py:72]:
```suggestion
    """Converts a JSON schema property to a `pydantic.Field`.
```

File: src/distilabel/steps/tasks/structured_outputs/utils.py [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+
+    Args:
+        name: The field name.
+        json_schema: The JSON schema property.
+
+    Returns:
+        A Pydantic field definition.
```
Comment (gabrielmbmb) [inline src/distilabel/steps/tasks/structured_outputs/utils.py:79]:
```suggestion
        A `pydantic.Field`.
```

File: src/distilabel/steps/tasks/structured_outputs/utils.py [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+        properties = json_schema.get("properties")
+        if properties:
+            nested_model = json_schema_to_model(json_schema)
+            return nested_model
+        else:
+            return Dict
+    elif type_ == "null":
+        return Optional[Any]  # Use Optional[Any] for nullable fields
+    else:
+        raise ValueError(f"Unsupported JSON schema type: {type_}")
```
Comment (gabrielmbmb) [inline src/distilabel/steps/tasks/structured_outputs/utils.py:153]: not big fan of `elif`

File: src/distilabel/steps/tasks/structured_outputs/utils.py (hunk 3 above)
Comment (gabrielmbmb) [reply src/distilabel/steps/tasks/structured_outputs/utils.py:38]: missing args and return docstring

File: docs/sections/learn/advanced/structured_generation.md [hunk 9]
```diff
@@ -119,6 +127,72 @@ These were some simple examples, but one can see the options this opens.
     curl -L -o ~/Downloads/openhermes-2.5-mistral-7b.Q4_K_M.gguf https://huggingface.co/TheBloke/OpenHermes-2.5-Mistral-7B-GGUF/resolve/main/openhermes-2.5-mistral-7b.Q4_K_M.gguf
     ```
 
+## Instructor
+
+When working with model providers behind an API, there's no direct way of accesing the internal logit processor as `outlines` does, but thanks to [`instructor`](https://python.useinstructor.com/) we can generate structured output from LLM providers. We have integrated `instructor` to deal with the [`AsyncLLM`][distilabel.llms.AsyncLLM], so you can work with the following LLMs: [`OpenAILLM`][distilabel.llms.OpenAILLM], [`AzureOpenAILLM`][distilabel.llms.AzureOpenAILLM], [`CohereLLM`][distilabel.llms.CohereLLM], [`GroqLLM`][distilabel.llms.GroqLLM], [`LiteLLM`][distilabel.llms.LiteLLM] and [`MistralLLM`][distilabel.llms.MistralLLM].
```
[the remark this answers is not in the document]
Comment (gabrielmbmb) [reply docs/sections/learn/advanced/structured_generation.md:132]: after reviewing the code, maybe it's OK to leave `AsyncLLM`.

File: docs/sections/pipeline_samples/examples/index.md
[the remark this answers is not in the document]
Comment (plaguss) [reply docs/sections/pipeline_samples/examples/index.md:1]: Do you mean creating a specific subpage within the `/examples` in the docs?

File: src/distilabel/llms/base.py [hunk 10]
```diff
@@ -300,3 +303,83 @@ def __del__(self) -> None:
...
+            variable in case it is a pydantic model.
+        """
+        from distilabel.steps.tasks.structured_outputs.instructor import (
+            prepare_instructor,
+        )
+
+        result = {}
+        client = prepare_instructor(
+            client,
+            mode=structured_output.get("mode"),
```
[the remark this answers is not in the document]
Comment (plaguss) [reply src/distilabel/llms/base.py:332]: We are in the second case indeed, so I will remove the `Optional` type here

**Review by alvarobartt — APPROVED** [review]

**Pull request merged**